### DISCUSSION TITLE
Custom self-set version of package?

### DISCUSSION THREAD

**Discussion by ognjen-andric** [projen/projen#2429] [Q&A] [answered]
Hey all,

Trying to figure out how to appropriately set our package versions while developing shared CDK constructs.
We have two different environments for shared libraries, preprod and prod. Preprod is being used while developing and production is the environment where our released libraries should end up at.
Each environment has its own npm repository which it uses.

While developing, we are using another project as a consumer project to test if everything works as intended.
**Is there a possibility to set custom versions without relying on Git tags?**

We wish to publish tags only for production releases (versions that are used by other people), and to release custom versions while developing onto our private npm repository without pushing new tags on our Git.

For example, ideally would be : 
Production : `v1.0.5`
Preprod : `feat-v1.0.5-<ticketNumber>-<releaseVersion>`

So while developing, every time we push new version onto our feature branch, new version would be released and it would look like : 
- `feat-v1.0.5-ticket123-0`
- `feat-v1.0.5-ticket123-1`
- `feat-v1.0.5-ticket123-2`

This would allow us not to "dirty" our Git until we release to production and to allow us to increment versions uniquely per feature and ticket number to make sure our consumer projects can use them if needed.

Any recommendation on how to do this (or if it is even possible)?

**Comment by ognjen-andric** [comment]
Ideally if we could somehow set a version for next release in `.projenrc.js` file or set a flag while releasing the package, this would enable us to write our logic to behave differently for production and preproduction. However, I wasn't lucky to find solution to this while reading through documentation (or might have missed it)

**Comment by mrgrain** [comment, marked as answer]
projen's release system is definitely not setup for this. 😅  But there are ways to work around it.

This will set a version to an explicit string:
```ts
project.package.addVersion('1.2.3-banana');
```

Now build _should_ give you a releaseable artifact with the correct version number, somewhere in `dist`:
```console
projen build
```

Finally you need to adjust the release workflows and/or tasks to only do something like 
```console
npm publish dist/js/my-package-1.2.3-banana.tgz
```

Let me know if that works for you!

**Comment by mvs5465** [comment]
This is what we do in the `.projenrc.js` file, not sure if its different than the above

```
project.addFields({ version: '0.0.152' });
```
